### ISSUE TITLE
Implement dodge by shifting each category if necessary, based on the dodge vector

### ISSUE THREAD

**Issue by asinghvi17** [asinghvi17/SwarmMakie.jl#7]
With all this handling, maybe we should move the category code into the recipe?  But then there would be no control...

**Comment by TheCedarPrince** [comment]
To make sure I am understanding what you are saying with dodging, you are imagining a sort of dodging figure like this:

![image](https://github.com/asinghvi17/SwarmMakie.jl/assets/29561456/662db38b-c7b2-422c-806a-209943227882)


Out of curiosity, what is the process you are imagining? I am struggling to imagine what the API would look like with:

```julia
beeswarm(categories, ys, dodge = ?)
```

**Comment by asinghvi17** [comment]
Yep, that's what I'd envision.  Dodging would mutate the xs of position before applying calculate.  Similar to bar/violin plots.

**Comment by TheCedarPrince** [comment]
Gotcha -- how would we want to handle dodge grouping? Would we want to just have an API like Makie's boxplot:

```julia
using CairoMakie

categories = rand(1:3, 1000)
values = randn(1000)
dodge = rand(1:2, 1000)

boxplot(categories, values, dodge = dodge, color = dodge)
```

Or were you thinking of something else?

**Comment by asinghvi17** [comment]
Nope, it would be pretty much that.  Maybe we could hoist the category utilization code out of `calculate!` and into the main `onany` block...

**Comment by TheCedarPrince** [comment]
Brilliant! Last question -- how would we want to `gutterize!` dodged groups? Would there be the typical gutter per category overall but then also like an `inside_gutter` dimension that would set thresholds inside the category between dodged groups? 

What do you think @asinghvi17 ? Either way, I'll work on a naive implementation to see what I can do here. I like the idea of hoisting things by the way. But maybe, let's wait until we have the feature fleshed out some more before we do a bit more of an overhaul?

**Comment by asinghvi17** [comment]
I would imagine so - but that could also be based on the dodge width!

There is already a readymade utility function in Makie:
https://github.com/MakieOrg/Makie.jl/blob/14b65a0e8c786030f095cfe37860e6eaeec39eaa/src/basic_recipes/barplot.jl#L101-L115

and the way `violin` (probably the closest plot to a beeswarm) manages it is by using multiple attributes to control dodge.
https://github.com/MakieOrg/Makie.jl/blob/14b65a0e8c786030f095cfe37860e6eaeec39eaa/src/stats/violin.jl#L26-L29
